research/minigo/dualnet_model.py

```diff
@@ -0,0 +1,236 @@
+    """Initialize a model for computing the policy and value in RL.
+
+    Args:
+      num_filters: Number of filters (AlphaGoZero used 256). We use 128 by
+        default for a 19x19 go board, and 32 for 9x9 size.
+      num_shared_layers: Number of shared residual blocks.  AGZ used both 19
+        and 39. Here we use 19 for 19x19 size and 9 for 9x9 size because it's
+        faster to train.
+      fc_width: Dimensionality of the fully connected linear layer.
+      go_size: A single integer for the board size.
+    """
+    self.num_filters = num_filters
+    self.num_shared_layers = num_shared_layers
+    self.fc_width = fc_width
+    self.go_size = go_size
+    self.kernel_size = [3, 3]  # kernel size is from AGZ paper
+
+  def __call__(self, inputs, training):
+    """Add operations to classify a batch of input Go features.
+
+    Args:
+      inputs: A Tensor representing a batch of input Go features with shape
+        [BATCH_SIZE, go_size, go_size, features.NEW_FEATURES_PLANES]
+      training: A boolean. Set to True to add operations required only when
+        training the classifier.
+
+    Returns:
+      policy_logits: A vector of size self.go_size * self.go_size + 1
+        corresponding to the policy logit probabilities for all intersections
+        and the pass move.
+      value_logits: A scalar for the value logits output
+    """
+    initial_output = _conv_block(
+        inputs=inputs, filters=self.num_filters,
+        kernel_size=self.kernel_size, training=training)
+    # the shared stack
+    shared_output = initial_output
+    for _ in range(self.num_shared_layers):
+      shared_output = _res_block(
+          inputs=shared_output, filters=self.num_filters,
+          kernel_size=self.kernel_size, training=training)
+
+    # policy head
+    policy_conv2d = _conv2d(inputs=shared_output, filters=2, kernel_size=[1, 1])
+    policy_batchn = _batch_norm(inputs=policy_conv2d, training=training,
+                                center=False, scale=False)
+    policy_relu = tf.nn.relu(policy_batchn)
+    policy_logits = tf.layers.dense(
+        tf.reshape(policy_relu, [-1, self.go_size * self.go_size * 2]),
+        self.go_size * self.go_size + 1)
+
+    # value head
+    value_conv2d = _conv2d(shared_output, filters=1, kernel_size=[1, 1])
+    value_batchn = _batch_norm(value_conv2d, training,
+                               center=False, scale=False)
+    value_relu = tf.nn.relu(value_batchn)
+    value_fc_hidden = tf.nn.relu(tf.layers.dense(
+        tf.reshape(value_relu, [-1, self.go_size * self.go_size]),
+        self.fc_width))
+    value_logits = tf.reshape(tf.layers.dense(value_fc_hidden, 1), [-1])
+
+    return policy_logits, value_logits
+
+
+def model_fn(features, labels, mode, params, config=None):  # pylint: disable=unused-argument
+  """DualNet model function.
+
+  Args:
+    features: tensor with shape
+      [BATCH_SIZE, self.go_size, self.go_size, features.NEW_FEATURES_PLANES]
```

Comment:
nit: we seem to use go.N, go_size, and self.go_size interchangeably here. We should pick a name and stick with it, esp for this case, where there is no self for the model_fn.